src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -832,7 +833,26 @@ abstract class RefChecks extends Transform {
           }
           member resetFlag (OVERRIDE | ABSOVERRIDE)  // Any Override
         }
-    }
+
+      // 5. Check that the nested class do not shadow other nested classes from outer class's parent
+      def checkNestedClassShadow(): Unit =
```

Comment:
In principle, `clazz.allOverriddenSymbols.filter(_.isClass)` would be the simplest solution, but this condition rules it out:

https://github.com/scala/scala/blob/57f2e5700927808c8f0dd547d9890c833d5d5900/src/reflect/scala/reflect/internal/Symbols.scala#L2424-L2428

Inlining through the implementation of `allOverriddenSymbols`, this seems to work:

```
scala> :power
Power mode enabled. :phase is at typer.
import scala.tools.nsc._, intp.global._, definitions._
Try :help or completions for vals._ and power._

scala> class C { class I }
defined class C

scala> class D extends C { class I }
defined class D

scala> val clazz = typeOf[D].member(TypeName("I"))
clazz: $r.intp.global.Symbol = class I

scala> val os = clazz.owner.ancestors.map(a => clazz.matchingSymbol(a, clazz.owner.thisType)).filter(c => c.exists && c.isClass)
os: List[$r.intp.global.Symbol] = List(class I)

scala> os.map(_.fullNameString)
res14: List[String] = List(C.I)
```